Prometheus metrics for measuring requests and responses in Proxy Node

idrop opened the pull request:
🆕 _tested and working in sandbox cluster_

Generate metrics for Prometheus that count:
- requests into Proxy Node gateway
- successfully handled requests into Proxy Node gateway
- responses to Proxy Node gateway
- successfully handled responses to Proxy Node gateway

These metrics can be used for dashboards/alerts in Grafana.

Adding the `release` as a label allows filtering of these metrics by country and environment i.e.  `nl`, `it`, `it-integration` etc.
A useful side effect of this new label is that it it [can be used as a selector.](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/)
All services therefore have this label added.

In addition, the `translator` and `esp` are configured to provide out-of-the-box dropwizard metrics to Prometheus.

All releases within the Verify cluster are enabled.

For reference, [DCS merged a similar PR](https://github.com/alphagov/doc-checking/commit/a470142365202b499503bf3f3a9fc118faa24c90).

idrop committed:
Expose default Prometheus metrics plus counters to measure requests and responses into gateway

idrop commented on eidas-saml-parser/src/main/java/uk/gov/ida/notification/eidassaml/EidasSamlResource.java:
```diff
@@ -29,6 +30,12 @@
 @Produces(MediaType.APPLICATION_JSON)
 public class EidasSamlResource {
 
+    private static final Counter SUCCESSFUL_REQUESTS = Counter.build(
```
Agree, will move

philandstuff commented on eidas-saml-parser/src/dist/config.yml:
```diff
@@ -30,3 +30,5 @@ replayChecker:
 
 proxyNodeAuthnRequestUrl: ${PROXY_NODE_AUTHN_REQUEST_ENDPOINT:-DefaultNotUsed}
+
+prometheusEnabled: ${PROMETHEUS_ENABLED:-true}
```
do you anticipate ever switching this off? If not I'd hardcode it to `true` in code. hot take: configuration options are tech debt

philandstuff commented on eidas-saml-parser/src/main/java/uk/gov/ida/notification/eidassaml/EidasSamlParserConfiguration.java:
```diff
@@ -36,4 +40,9 @@ public TrustStoreBackedMetadataConfiguration getConnectorMetadataConfiguration()
+
+    @Override
+    public boolean isPrometheusEnabled() {
+        return prometheusEnabled;
```
```suggestion
        return true;
```
related to above comment

philandstuff commented on eidas-saml-parser/src/main/java/uk/gov/ida/notification/eidassaml/EidasSamlResource.java:
```diff
@@ -29,6 +30,12 @@
+    private static final Counter SUCCESSFUL_REQUESTS = Counter.build(
+            "verify_proxy_node_requests",
+            "Total number of successful Verify Proxy Node Requests")
+            .labelNames("issuer")
```
do we know the list of issuers at app start-up? If so, it's good practice to "initialise" the counter with the list of possible `issuer` label values by calling `SUCCESSFUL_REQUESTS.labels(issuer);` for each `issuer` value.

The cost of not doing this is that you miss the first instance of each request in your counters on each app startup.

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/GatewayConfiguration.java:
```diff
@@ -62,4 +65,9 @@ public URI getErrorPageRedirectUrl() {
+
+    @Override
+    public boolean isPrometheusEnabled() {
+        return prometheusEnabled;
```
```suggestion
        return true;
```

philandstuff commented on eidas-saml-parser/src/main/java/uk/gov/ida/notification/eidassaml/EidasSamlResource.java:
```diff
@@ -61,6 +68,7 @@ public EidasSamlParserResponse post(@Valid EidasSamlParserRequest request) throw
         ProxyNodeLogger.addContext(ProxyNodeMDCKey.EIDAS_ISSUE_INSTANT, authnRequest.getIssueInstant().toString());
         ProxyNodeLogger.addContext(ProxyNodeMDCKey.EIDAS_LOA, getLevelsOfAssurance(authnRequest));
         ProxyNodeLogger.info("Authn request validated by ESP");
+        SUCCESSFUL_REQUESTS.labels(authnRequest.getIssuer().getValue()).inc();
```
Do we want to count total requests as well as successful requests? That way, you can compute a rate of successes.

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/resources/HubResponseResource.java:
```diff
@@ -28,6 +30,12 @@
+            "Total number of successful Verify Proxy Node Responses")
+            .labelNames("destination")
+            .register();
+
```
Same comment as above re initialising counter for each label if possible

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/resources/HubResponseResource.java:
```diff
@@ -65,11 +73,13 @@ public View hubResponse(
+                sessionData.getEidasRelayState()
         );
+        SUCCESSFUL_RESPONSES.labels(sessionData.getEidasDestination()).inc();
+        return samlFormView;
```
same comment as above re counting total responses as well as successful responses

philandstuff commented on proxy-node-translator/src/main/java/uk/gov/ida/notification/translator/configuration/TranslatorConfiguration.java:
```diff
@@ -51,4 +55,9 @@ public CredentialConfiguration getCredentialConfiguration() {
+
+    @Override
+    public boolean isPrometheusEnabled() {
+        return prometheusEnabled;
```
```suggestion
        return true;
```

philandstuff reviewed (dismissed):
Looks like a good start. Left a bunch of comments, none of them are blockers.  Happy to have a chat face to face as well.

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/Metrics.java:
```diff
@@ -0,0 +1,51 @@
+    private static final String LABEL_PREFIX = "verify_proxy_node";
+
+    public static final Counter REQUESTS = Counter.build(
+            LABEL_PREFIX + "_requests",
```
```suggestion
            LABEL_PREFIX + "_requests_total",
```
The `_total` suffix is a [Prometheus naming convention](https://prometheus.io/docs/practices/naming/) that is relatively strongly enforced and tools like `promtool check metrics` or Grafana's explore mode will complain if Counters don't have `_total` (or if non-Counter metrics do have it).

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/Metrics.java:
```diff
@@ -0,0 +1,51 @@
+            .register();
+
+    public static final Counter REQUESTS_SUCCESSFUL = Counter.build(
+            LABEL_PREFIX + "_requests_successful",
```
```suggestion
            LABEL_PREFIX + "_successful_requests_total",
```

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/Metrics.java:
```diff
@@ -0,0 +1,51 @@
+            .register();
+
+    public static final Counter RESPONSES = Counter.build(
+            LABEL_PREFIX + "_responses",
```
```suggestion
            LABEL_PREFIX + "_responses_total",
```

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/Metrics.java:
```diff
@@ -0,0 +1,51 @@
+            .register();
+
+    public static final Counter RESPONSES_SUCCESSFUL = Counter.build(
+            LABEL_PREFIX + "_responses_successful",
```
```suggestion
            LABEL_PREFIX + "_successful_responses_total",
```

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/Metrics.java:
```diff
@@ -0,0 +1,51 @@
+
+    public static final String RELEASE_NAME = getReleaseName();
+
+    private static final String RELEASE_NAME_LABEL = "release";
```
I feel like this would be better added as a target label in the `ServiceMonitor` resource.  In general it's a Prometheus antipattern for an instance to need to know too much about its own identity; it should just be "a proxy-node instance" and the Prometheus scrape rules (in the ServiceMonitor) should add the context about exactly which proxy-node instance this is, in the form of target labels.

That also means you can get rid of a lot of the wiring around passing the RELEASE_NAME into the app as an envvar.

philandstuff commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/Metrics.java:
```diff
@@ -0,0 +1,51 @@
+            .register();
+
+    static { // initialise each label so first increment is not missed
+        REQUESTS.labels(RELEASE_NAME_LABEL);
```
For `Counter`s which have no labels, this is not required - the Counter is automatically initialized.  This is another argument in favour of moving the `release` label to the ServiceMonitor config.

philandstuff reviewed (commented):
some comments but we'll discuss in person

philandstuff committed:
remove `release` label from Gateway

Going to re-implement as a target label instead

philandstuff committed:
add ServiceMonitor to actually scrape the new metrics endpoints

This adds a ServiceMonitor.  What it does:

 - looks for Service instances based on a label selector
 - creates prometheus scrape targets for the pods within that Service,
   based on the `mgmt` port
 - adds the `release` label as a target label (this was removed from
   the apps themselves in the previous commit)

philandstuff committed:
fix metric names and descriptions

Counters should have the `_total` suffix:
https://prometheus.io/docs/practices/naming/

Tools like `promtool check http` and grafana's explore mode will
complain if counters don't adhere to this convention.

Also, don't use Title Case, use sentence case.  I couldn't find a
brilliant reference for this but the design system typography guide
says that headings should use sentence case:
https://design-system.service.gov.uk/styles/typography/#headings and I
think that our metric descriptions should match this.

philandstuff reviewed (dismissed):
I would approve this wouldn't I 😄

philandstuff committed:
fix it so it actually compiles

idrop committed:
Change name of class Metrics to MetricsUtils, in line with configured PMD rules checked by Codacy, see https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#classnamingconventions

idrop committed:
Label and selector values should be strings. Validated on sandbox.

galund commented on proxy-node-gateway/src/main/java/uk/gov/ida/notification/resources/EidasAuthnRequestResource.java:
```diff
@@ -96,7 +98,7 @@ private View handleAuthnRequest(String encodedEidasAuthnRequest, String eidasRel
         ProxyNodeLogger.addContext(ProxyNodeMDCKey.HUB_URL, vspResponse.getSsoLocation().toString());
         ProxyNodeLogger.info("Authn requests received from ESP and VSP");
-
+        MetricsUtils.REQUESTS_SUCCESSFUL.inc();
```
I don't know if this is a sensible complaint, but it feels like there is still some opportunity for the request to fail, after this counter has been incremented.

Perhaps we could put the value from `buildSamlFormView` into a variable, and increment the counter immediately before we `return`?

idrop replied:
Good catch, I've pushed to inc() the counter at last possible point

idrop committed:
Record successful request to gateway at last possible point
